render/naga-agal/src/builder.rs

```diff
@@ -808,43 +808,115 @@ impl<'a> NagaBuilder<'a> {
-                    self.evaluate_expr(Expression::Load { pointer: temp }),
+        let mut load_register = |register_type: &RegisterType, reg_num| {
+            match register_type {
+                // We can use a function argument directly - we don't need
+                // a separate Expression::Load
+                RegisterType::Attribute => Ok((
+                    self.get_vertex_input(reg_num as usize)?,
+                    self.shader_config.vertex_attributes[reg_num as usize]
+                        .ok_or(Error::MissingVertexAttributeData(reg_num as usize))?,
+                )),
+                RegisterType::Varying => Ok((
+                    self.emit_varying_load(reg_num as usize)?,
                     VertexAttributeFormat::Float4,
-                )
-            }
-            _ => {
-                return Err(Error::Unimplemented(format!(
+                )),
+                RegisterType::Constant => Ok((
+                    self.emit_const_register_load(reg_num as usize)?,
+                    // Constants are always a vec4<f32>
+                    VertexAttributeFormat::Float4,
+                )),
+                RegisterType::Temporary => Ok({
+                    let temp = self.get_temporary_register(reg_num as usize)?;
+                    (
+                        self.evaluate_expr(Expression::Load { pointer: temp }),
+                        VertexAttributeFormat::Float4,
+                    )
+                }),
+                _ => Err(Error::Unimplemented(format!(
                     "Unimplemented source reg type {:?}",
                     source.register_type
-                )))
+                ))),
             }
         };
 
-        if matches!(source.direct_mode, DirectMode::Indirect) {
-            return Err(Error::Unimplemented(
-                "Indirect addressing not implemented".to_string(),
-            ));
-        }
+        let (mut base_expr, source_type) = match source.direct_mode {
+            DirectMode::Direct => load_register(&source.register_type, source.reg_num)?,
+            DirectMode::Indirect => {
+                // Handle an indirect register load, e.g. `vc[va0.x + offset]`
+                // Indirect loads allow loading from a dynamically computed register index.
+                // This dynamic index is computed as 'regN.X + offset', where 'regN' is a normal
+                // register (e.g. 'va0'), and 'X' is a component ('X, 'Y', Z', or 'W').
+                // Currently, we only support this when the 'outer' (non-index) register is
+                // a constant register, since we always access constant registers through
+                // an array access.
+                match source.register_type {
+                    RegisterType::Constant => {
+                        // Load the index register (e.g. 'va0') as normal, and access the component
+                        // given by 'index_select'. This is 'va0.x' in the above example.
```

Comment:
(Very minor nit:

Replace:

```
// given by 'index_select'. This is 'va0.x' in the above example.
```

with

```
// given by 'index_select' (e.g. 'x'). This is 'va0.x' in the above example.
```

?)

Reply:
Good idea - I've changed it